src/numerics/funcs.cpp

```diff
@@ -27,4 +28,23 @@ doublereal linearInterp(doublereal x, const vector_fp& xpts,
     return ff;
 }
 
+double trapezoidal(const Eigen::VectorXd& f, const Eigen::VectorXd& x)
+{
+    // check length
+    if (f.size() != x.size()) {
+        throw CanteraError("trapezoidal",
+                           "Vector lengths need to be the same.");
+    }
+    double sum = 0;
+    for (auto i = 0; i < f.size() - 1; i++) {
+        if (x(i+1) - x(i) > 0.0) {
+            sum += (f(i+1) +  f(i)) / 2.0 * (x(i+1) - x(i));
+        } else {
+            throw CanteraError("trapezoidal",
+                "x (coordinate) needs to be the monotonically increasing.");
+        }
+    }
+    return sum;
+}
```

Comment:
I wonder if there's a way to vectorize this implementation, rather than using a loop. You can relatively easily compute the `diff` of the x vector (https://stackoverflow.com/a/42854359) and check that all the elements are positive. You can also possibly use convolution with a window of length 2 to compute the rolling mean of the `f` vector. Maybe it won't actually help anything, but my NumPy brain saw this and went 💡

Reply:
Good suggestion. I also implemented Eigen::LinSpaced in PlasmaPhase's initialtion.